File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        ans=[]
        nums.sort()
        for i in range(n-2):
            if(i>0 and nums[i]==nums[i-1]):
                continue
            if(nums[i]>0):
                break    
            sum=-1*nums[i]
            j=i+1
            k=n-1
            while(j<k):
                res=nums[j]+nums[k]
                if res==sum:
                    ans.append([nums[i],nums[j],nums[k]])
                    j+=1
                    k-=1
                    while(j<k and nums[j]==nums[j-1]):
                        j+=1
                    while(j<k and nums[k]==nums[k+1]):
                        k-=1    
                elif(res<sum):
                    j+=1
                else:
                    k-=1
        return ans            

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.cpp
class Solution {
public: int n;
    
    int f(int i,vector<int>&nums,vector<vector<int>>&dp,int sum){
        if(i>=n){
            if(sum==0) return 1;
            else{
                return 0;
            }
        }
        if(dp[i][sum]!=-1) return dp[i][sum];
        int nt=f(i+1,nums,dp,sum);
        int take=0;
        if(sum>=nums[i])
          take=f(i+1,nums,dp,sum-nums[i]);
        return dp[i][sum]= nt+take;
    }

    bool canPartition(vector<int>& nums) {
        int sum=0; n=nums.size();
        for(int i=0;i<n;i++){
            sum+=nums[i];
        }
        if(sum%2==1) return false;
        vector<vector<int>>dp(n+1,vector<int>(sum/2+1,-1));
        return f(0,nums,dp,sum/2);
        
    }
};
